### DISCUSSION TITLE
validate form

### DISCUSSION THREAD

**Discussion by MohamadrezaKazemian** [prisma/prisma#17060] [Q&A] [answered]
Hi, I'm new to Prisma. my API looks like this : 
```

import type { NextApiRequest, NextApiResponse } from 'next';
import prisma from '../../../lib/prisma';

export default async (req: NextApiRequest, res: NextApiResponse) => {
    if (req.method !== 'POST') {
        return res.status(405).json({ message: 'Method not allowed' });
    }
    try {
        const data = JSON.parse(req.body);
        const saveProfile = await prisma.businessOwners.updateMany({
            where:{ 
                id : data.id
            },
            data:{
                email : data.email,
                fullName : data.fullName,
                password : data.password,
                phone : data.phone,
                username : data.username,
            }
        })
        res.status(200).json(saveProfile);
    } catch (err) {
        res.status(400).json({ message: `${err}` });
    }
};

```
I want to validate my API form so that it would be more secure. but how can I make that happen? 
for example, checking if the email is really email or the password is between 5 to 10 characters does my code look fine right now?
thank you for your help developers!

p.s : I'm using Nextjs + Prisma

**Comment by nurul3101** [comment, marked as answer]
Hello @MohamadrezaKazemian 👋 

You can use libraries like [Joi](https://joi.dev/api/?v=17.7.0) or [Zod](https://zod.dev/?id=strings) to validate your data. Joi and Zod has inbuilt functions like .email() to validate if a string is a valid email address. You can use the min() and max() functions of these libraries to validate the length of the password field.

Also, if you don't want to use these libraries you can use a regex to validate if the provided string is an email and also to check the length of the password.

It would look something like this:
```typescript
        // Validate email
        if (!/^\S+@\S+$/.test(data.email)) {
            return res.status(400).json({ message: 'Invalid email' });
        }

        // Validate password
        if (data.password.length < 5 || data.password.length > 10) {
            return res.status(400).json({ message: 'Password must be between 5 and 10 characters' });
        }
```

**Reply by MohamadrezaKazemian** [reply]
thank you for your response.
 could you please show me one example of doing validation with zod or joi with my source code? 
i am using the regex and length but frankly im getting tierd of it!

**Reply by nurul3101** [reply]
It could look something like this with Joi:

```typescript
import type { NextApiRequest, NextApiResponse } from 'next';
import prisma from '../../../lib/prisma';
import Joi from 'joi';

const schema = Joi.object().keys({
  email: Joi.string().email().required(),
  fullName: Joi.string().required(),
  password: Joi.string().min(5).max(10).required(),
  phone: Joi.string().required(),
  username: Joi.string().required(),
});

export default async (req: NextApiRequest, res: NextApiResponse) => {
  if (req.method !== 'POST') {
    return res.status(405).json({ message: 'Method not allowed' });
  }
  try {
    const data = JSON.parse(req.body);

    // Validate the form data using the schema
    const { error, value } = schema.validate(data);
    if (error) {
      return res.status(400).json({ message: error.message });
    }

    // Save the profile
    const saveProfile = await prisma.businessOwners.updateMany({
      where: {
        id: data.id,
      },
      data: {
        email: data.email,
        fullName: data.fullName,
        password: data.password,
        phone: data.phone,
        username: data.username,
      },
    });
    res.status(200).json(saveProfile);
  } catch (err) {
    res.status(400).json({ message: `${err}` });
  }
};
```

**Reply by MohamadrezaKazemian** [reply]
you are a life saver! thank you very much for your time and helps!!
the only question remaining for me is that is it the right way to validate my api? i mean is it clean and secure? for a real project on production

**Reply by nurul3101** [reply]
Yes, using Joi or Zod to validate your data is an ideal approach and it is used in numerous production projects, so you should not face any issues in using them for your production project.

**Reply by MohamadrezaKazemian** [reply]
thank you very much for your help <3

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
